src/lib/storage/segment_access_statistics.hpp

```diff
@@ -0,0 +1,379 @@
+#include <string>
+
+#include <types.hpp>
+#include <storage/pos_list.hpp>
+
+namespace opossum {
+class Table;
+// -----------------------------------------------------------------------------------------------------------------
+enum class SegmentAccessType {
+  Other,
+  IteratorCreate,
+  IteratorSeqAccess,
+  IteratorIncreasingAccess,
+  IteratorRandomAccess,
+  IteratorDereference,
+  AccessorCreate,
+  AccessorAccess,
+  DictionaryAccess,
+  // number of elements in enum
+    Count
+};
+
+// -----------------------------------------------------------------------------------------------------------------
+enum SegmentType {
+  Dictionary,
+  FrameOfReference,
+  LZ4,
+  Reference,
+  RunLength,
+  ValueS
+};
+
+// -----------------------------------------------------------------------------------------------------------------
+
+class SegmentAccessTypeTools {
+ public:
+  inline static const std::string headers = "Other,IteratorCreate,IteratorSeqAccess,IteratorIncreasingAccess,"
+                                            "IteratorRandomAccess,IteratorDereference,AccessorCreate,AccessorAccess,"
+                                            "DictionaryAccess";
+
+  static SegmentAccessType iterator_access_pattern(const std::shared_ptr<const PosList>& positions);
+
+ private:
+  // There are five states:
+  // 0 (unknown), equivalent to IteratorSeqAccess
+  // 1 (sequentially increasing), difference between two neighboring elements is 0 or 1.
+  // 2 (increasing randomly)
+  // 3 (sequentially decreasing), difference between two neighboring elements is -1 or 0.
+  // 4 (decreasing randomly)
+  // 5 (random access)
+  enum class State {Unknown, SeqInc, RndInc, SeqDec, RndDec, Rnd, Count};
+
+  // There are five possible inputs
+  enum class Input {Zero, One, Positive, NegativeOne, Negative, Count};
+
+  constexpr static const std::array<std::array<State, static_cast<uint32_t>(Input::Count)>, static_cast<uint32_t>(State::Count)>
+    _transitions{{{State::Unknown, State::SeqInc, State::SeqInc, State::SeqDec, State::SeqDec},
+                  {State::SeqInc, State::SeqInc, State::RndInc, State::Rnd, State::Rnd},
+                  {State::RndInc, State::RndInc, State::RndInc, State::Rnd, State::Rnd},
+                  {State::SeqDec, State::Rnd, State::Rnd, State::SeqDec, State::RndDec},
+                  {State::RndDec, State::Rnd, State::Rnd, State::RndDec, State::RndDec},
+                  {State::Rnd, State::Rnd, State::Rnd, State::Rnd, State::Rnd}}};
+};
+// -----------------------------------------------------------------------------------------------------------------
+class AtomicAccessStrategy {
+ public:
+  uint64_t count(SegmentAccessType type) const;
+
+  void reset();
+
+  void increase(SegmentAccessType type, uint64_t count);
+
+  static std::string header();
+
+  std::vector<std::string> to_string() const;
+
+ private:
+  std::array<std::atomic_uint64_t, static_cast<uint32_t>(SegmentAccessType::Count)> _count;
```

Comment:
Took me a while to understand that "Count" is not an access type but a magic value. Have you considered magic_enum?

Reply:
That is a great idea.
Moving away from the vector and using a struct instead might be a better alternative in the long run though.